corehq/apps/es/forms.py

```diff
@@ -57,9 +69,92 @@ class ElasticForm(ElasticDocumentAdapter):
     def mapping(self):
         return get_adapter_mapping(self)
 
-    @classmethod
-    def from_python(cls, doc):
-        return from_dict_with_possible_id(doc)
+    def from_python(self, xform):
+        from corehq.form_processor.models.forms import XFormInstance
+        if isinstance(xform, dict):
+            xform_dict = deepcopy(xform)
+        elif isinstance(xform, XFormInstance):
+            xform_dict = xform.to_json()
+        else:
+            raise TypeError(f"Unknown type {type(xform)}")
+        return self.from_dict(xform_dict)
+
+    def from_dict(cls, xform):
+        """
+        Takes in a xform dict and applies required transformation to make it suitable for ES.
+        The function is replica of ``transform_form_for_elasticsearch``
+        and will be replaced by `from_python` in future
+
+        :param xform: an instance of ``dict`` which is ``XFormInstance.to_json()``
+        """
+        from casexml.apps.case.xform import extract_case_blocks
+
+        from corehq.apps.receiverwrapper.util import get_app_version_info
+        from corehq.pillows.utils import format_form_meta_for_es, get_user_type
+        from corehq.pillows.xform import is_valid_date
+
+        doc_ret = deepcopy(xform)
```

Comment:
Is this one just too hard to eliminate `deepcopy`? In terms of runtime cost, it may still be worth it even if it's tedious.

Reply:
[4b9d19a](https://github.com/dimagi/commcare-hq/pull/32660/commits/4b9d19ae0333c1476dde2736f0cad8efe9139ee5)